cohttp-eio: add cochttp-eio based benchmark

bikallem opened the pull request:
Change summary:
1. Creates a dockerfile for each web server so that we can individually build/run benchmark as required. 
2. Add Makefile target `build` and `run` to to build and run separately
3. Add multi core benchmark script separate from 1 core. 
4. Add multi core rust benchmark.

Commit:
cohttp-eio: add pure eio ocaml-cohttp benchmark

vendors `eio` and `ocaml-cohttp#eio-3`.

Add cohttp_eio.exe for benchmark inclusion.

updates benchmark scripts

talex5 reviewed (changes requested):
Since this implementation is giving impossible results, some changes are obviously needed.

I suggest writing a validator program that behaves a bit like wrk2 (i.e. sends lots of requests in parallel) but ensures that it's getting correct responses. In particular:

1. Make sure it gets exactly the expected reply to each request. No truncated or extended versions.
2. Make sure it never gets more replies than it asked for.
3. Check that all connections are being serviced. It's quite easy to game the benchmark by e.g. just sending replies on 10 connections really fast, while stalling all other connections.

That should show what's wrong with your code, and would also be a really useful check that the others are working correctly (before then testing their speed with `wrk2`, as before).

bikallem commented:
With regards to the correctness of cohttp-eio, here is a small demo of it working. The response size, and the response in general seems to be correct. Below shows the server working along side the browser. 
 
![cohttp-eio](https://user-images.githubusercontent.com/315907/152015172-a1a85db6-0365-4eb9-9d94-03075ae1f0c6.gif)

~~After going through some literature regarding wrk2, it seems the wrk2 methodology is quite solid and tested in the field. At least the arguments for hdr histrogram and coordinated omissions sound very cogent. The response serviced/second is a sampling value, so I am inclined to think this is also correct(?). Perhaps https://github.com/gs0510 can confirm.~~

What puzzles me most is the steep drop in/around 150k/sec requests? Is there a memory space contention perhaps between the load generator and the server?

References:
1. https://github.com/giltene/wrk2 - Readme is quite comprehensive
2. https://www.quora.com/In-Java-what-is-Coordinated-Omission 
3. http://hdrhistogram.org - central to wrk2 methodology I believe

art-w commented:
So, this was "interesting". If you look at wrk2 log `output/run-cohttp_eio.exe-1000-1000.txt`:

```
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency    35.43s    14.26s    1.00m    57.73%
    Req/Sec    10.17k   390.77    10.63k    48.00%
...
  14978619 requests in 1.00m, 30.63GB read
```

How can wrk2 perform 15 million requests when it is scheduled to run a thousand per second, and how can the requests have an average of 35s latency?

Well, it turns out that wrk2 throttles the requests when *it* send them... but it doesn't throttle when reading the response from the server. As wrk2 is running in keep alive mode, the server is allowed to respond multiple times and wrk2 counts each additional response as a completed "request". However, wrk2 measures the response latency from the moment the initial http query was sent, so each new response from the server has an increasing latency (from 0ms for the first response to 1min for the last, yielding a large avg).

So yeah in practice, wrk2 sends 1'000 initial requests... and then get flooded with a never ending stream of responses.

I haven't had time to look into cohttp-eio, but you can reproduce the infinite stream of messages with:

```
$ # echo -e "GET / HTTP/1.1\r\nHost: localhost\r\n\r\n" | nc localhost 8080
HTTP/1.1 200 OK
Date: Thu, 03 Feb 2022 00:04:15 GMT
content-length: 2053
connection: keep-alive
content-type: text/plain; charset=UTF-8

CHAPTER I. Down the Rabbit-Hole  ......HTTP/1.1 200 OK
Date: Thu, 03 Feb 2022 00:04:15 GMT
content-length: 2053
connection: keep-alive
content-type: text/plain; charset=UTF-8

CHAPTER I. Down the Rabbit-Hole ......HTTP/1.1 200 OK
Date: Thu, 03 Feb 2022 00:04:15 GMT
content-length: 2053
connection: keep-alive
content-type: text/plain; charset=UTF-8

...
```

bikallem commented:
> I haven't had time to look into cohttp-eio, but you can reproduce the infinite stream of messages with:

I think I have pinpointed the issue. I will re-run the benchmark and post updates.

Commit:
One Dockerfile per benchmark

This commit creates one Dockerfile per benchmark. This enables
quickly and frequently running benchmarks when only one of the
benchmarks change.

If you want to rebuild an image, do the following,
make nethttp-go nc=--no-cache

Commit:
cohttp-eio: unvendor eio

Commit:
cohttp-eio: use eio from opam

Commit:
cohttp-eio: don't use uri

Commit:
cohttp-eio: use eio_linux and use polling

Commit:
cohttp-eio: use only half number of cores for httpaf-eio

Commit:
cohttp-eio: update to latest ocaml-cohttp

bikallem committed:
Merge branch 'master' into eio-3

Commit:
httpaf-eio: use latest docker script from master

Merges latest httpaf-eio build from master and adapts it.

bikallem commented:
This is ready now.

Commit:
cohttp-eio: update to latest ocaml-cohttp

Commit:
cohttp-eio: update Dockerfile

Commit:
rust-hyper: add multicore version of rust

Commit:
httpaf-eio: update to latest eio

Commit:
add script to run multi-core benchmarks

bikallem committed:
Merge branch 'master' into eio-3

bikallem commented:
@talex5 This is ready to merge.

talex5 commented on rust-hyper/src/main.rs:
```diff
@@ -2,54 +2,98 @@ extern crate futures;
 extern crate tokio_core;
 
-/*
-This is adapted from the Techempower Hyper benchmark: https://github.com/TechEmpower/FrameworkBenchmarks/tree/master/frameworks/Rust/hyper
```
This seems like useful information - why was it removed?

bikallem replied:
I think it got removed during code shuffling. Added again.

talex5 reviewed (commented):
Thanks - this looks reasonable now. What were the final results?

Commit:
add comment

bikallem commented:
Should I merge this if there aren't any objections?
